functorch/experimental/_scan.py

```diff
@@ -0,0 +1,370 @@
+        carry = pytree.tree_unflatten(flat_init, init_spec)
+        xs = pytree.tree_unflatten(flat_xs, xs_spec)
+        unflattened_carry_out, unflattened_out = f(carry, xs)
+        flat_carry_out, tmp_carry_out_spec = pytree.tree_flatten(unflattened_carry_out)
+        flat_out, tmp_out_spec = pytree.tree_flatten(unflattened_out)
+
+        nonlocal carry_out_spec
+        nonlocal out_spec
+        carry_out_spec = tmp_carry_out_spec
+        out_spec = tmp_out_spec
+        return *flat_carry_out, *flat_out
+    
+    flat_carry_out, flat_out = scan_impl(flat_fn, flat_init, flat_xs, return_all_carries=False, reverse=False)  # (ys, carry)
+    return pytree.tree_unflatten(flat_carry_out, carry_out_spec), pytree.tree_unflatten(flat_out, out_spec)
+
+class ScanAutogradOp(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, fw_graph, joint_graph, num_init_args, *flat_args):
+        ctx._joint_graph = joint_graph
+        ctx._num_input_args = len(flat_args)
+        ctx._num_init_args = num_init_args
+        with torch._C._AutoDispatchBelowAutograd():
+            flat_carry_out, flat_out, carries = scan_impl(fw_graph, flat_args[:num_init_args], flat_args[num_init_args:], return_all_carries=True, reverse=False)
+            # needs to flat the carries nested list
+            # second call to save_for_backward will override whatever that is saved earlier
+            # therefore, need to save everything that is needed at once
+            flat_carries = [carry_entry for carry in carries for carry_entry in carry]
+            # TODO: we don't need to actually store the init args, just need the xs args
+            ctx.save_for_backward(*flat_args, *flat_carries)
+            ctx._num_out_args = len(flat_out)
+            return (*flat_carry_out, *flat_out, *flat_carries)
+
+    @staticmethod
+    def backward(ctx, *flat_grads):
+        flat_carries = ctx.saved_tensors[ctx._num_input_args:]
+        fwd_args = ctx.saved_tensors[:ctx._num_input_args]
+        # bring back the nested carries list by chunking up the list to equal lengthed nested lists
+        # all carries should have length num_init_args
+        carries = [flat_carries[i*ctx._num_init_args:(i+1)*ctx._num_init_args] for i in range(len(flat_carries)//ctx._num_init_args)]
+        flat_xs = fwd_args[ctx._num_init_args:]
+        # carry and init should have the same spec
+        final_carry_grad = flat_grads[:ctx._num_init_args]
+        # ys should only be up to ctx._num_init_args + ctx._num_out_args, since the remaining should be grads for flat_carries
+        # and should be empty
+        ys_grad = flat_grads[ctx._num_init_args:ctx._num_init_args + ctx._num_out_args]
+        with torch._C._AutoDispatchBelowAutograd():
+            flat_carry_out_grads, flat_out_grads = scan_impl(ctx._joint_graph, final_carry_grad, (*ys_grad, carries, *flat_xs), return_all_carries=False, reverse=True)
```

Comment:
I don't think scan_impl needs a `return_all_carries` argument.

Let's say that the user wrote `scan(f, init, xs) -> carry, ys`. If autograd support is necessary, then with the "recompute" approach taken in this PR, we need to save all of the carries. 

We should be able to rewrite the user's scan into `scan_impl(f, init, xs) -> carry, (carries, ys)` and then just return (carry, ys).

Reply:
also might be some "checkpointing" methods :)